### PR TITLE
Add devilutionx.utils module

### PR DESCRIPTION
**Pull request by nelchael**
See discussion: <https://github.com/diasurgical/devilutionX/pull/6775#issuecomment-1793493004>.

### CODE REVIEW

**Comment by AJenbo** [comment]
This lib exposes some features I'm not sure are safe so it needs some consideration before/how we enable it.

https://www.lua.org/pil/22.html

image

**Comment by nelchael** [comment]
One way would be to always use bundled Lua and patch out the function, with this patch I get the following in the console:

Screenshot from 2023-11-04 19-28-55

File: 3rdParty/Lua/os-module-safety.patch [hunk 1]
```diff
@@ -0,0 +1,23 @@
+diff '--color=auto' -Nru Lua-88246d621abf7b6fba9332f49229d507f020e450.orig/lua-5.4.6/src/loslib.c Lua-88246d621abf7b6fba9332f49229d507f020e450/lua-5.4.6/src/loslib.c
+--- Lua-88246d621abf7b6fba9332f49229d507f020e450.orig/lua-5.4.6/src/loslib.c	2023-11-04 19:06:10.234901751 +0100
++++ Lua-88246d621abf7b6fba9332f49229d507f020e450/lua-5.4.6/src/loslib.c	2023-11-04 19:06:45.459865904 +0100
```
Comment (glebm) [inline 3rdParty/Lua/os-module-safety.patch:3]: We can't patch it like this -- the user might be using system Lua.
To do this safely, one can create a sandbox environment, and only copy the safe functions into the os module https://gitlab.com/OpenMW/openmw/-/blob/c1f7a9c258fdebaf677579eab3a513f3c7e7271a/components/lua/luastate.cpp#L272

**Review by glebm — CHANGES_REQUESTED** [review]

File: 3rdParty/Lua/os-module-safety.patch (hunk 1 above)
Comment (nelchael) [reply 3rdParty/Lua/os-module-safety.patch:3]: I've modified `CMake/Dependencies.cmake` to always use the bundled Lua.

Comment (glebm) [reply 3rdParty/Lua/os-module-safety.patch:3]: We'll want a sandboxed environment eventually anyway, so I don't think this is a good idea.
See also my second point (I edited the comment) about us potentially wanting to use it for internal (trusted) Lua modules.

Comment (glebm) [reply 3rdParty/Lua/os-module-safety.patch:3]: Here is the exact line from OpenMW that restricts the OS packages in the sandbox:

```
        mCommonPackages["os"] = mSandboxEnv["os"]
            = makeReadOnly(tableFromPairs<std::string_view, sol::function>({ { "date", mSol["os"]["date"] },
                { "difftime", mSol["os"]["difftime"] }, { "time", mSol["os"]["time"] } }));
```

Comment (nelchael) [reply 3rdParty/Lua/os-module-safety.patch:3]: Got it!

Comment (nelchael) [reply 3rdParty/Lua/os-module-safety.patch]: > We'll want a sandboxed environment eventually anyway, so I don't think this is a good idea. See also my second point (I edited the comment) about us potentially wanting to use it for internal (trusted) Lua modules.

It's out of scope for me for sure, I'd need to dig into how Lua internals work to take a stab at this, especially after briefly looking at the OpenMW code :(

Comment (glebm) [reply 3rdParty/Lua/os-module-safety.patch]: No worries, we'll get around to it eventually for sure!

**Comment by nelchael** [comment]
@AJenbo and @glebm -- as the `os` module may be problematic: any opinion on adding a new module, something like `devilutionx.util` with the equivalent of `date()`?

**Comment by AJenbo** [comment]
I'm not rejecting it, but I think we should mull over how to best do this. Glebm's suggestion [here](https://github.com/diasurgical/devilutionX/pull/6776#discussion_r1382443090) sounds interesting as it would mean we stay mostly compatible with standard Lua.

One thing is sure though, we will need some of the time related functions one way or another so the issue will be solved before we are ready for the next release.

**Comment by nelchael** [comment]
Agree, and `util` (or `utils`) may become the grab-bag of functions that do not fit anywhere else. The nice thing about a separate module is that it does not affect the sandbox environment implementation at all, and could actually simplify it a tiny bit by not needing to sanitize the `os` module (and make it outright unavailable).

**Comment by nelchael** [comment]
Ok, this is the implementation for `devilutionx.utils` module, right now it has only `date()` function which behaves like `os.date()`.

**Comment by glebm** [comment]
I've implemented the sandbox in https://github.com/diasurgical/devilutionX/pull/6778, it was easier than I thought

**Comment by nelchael** [comment]
> I've implemented the sandbox in #6778, it was easier than I thought

Nice! Should this PR be closed then?

**Comment by glebm** [comment]
Yeah no need for this one I think

**Pull request closed without merging**